metadata_service/proxy/neo4j_proxy.py

```diff
@@ -715,9 +719,136 @@ def get_user_detail(self, *, user_id: str) -> Union[UserEntity, None]:
+        RETURN COLLECT(DISTINCT tbl) as table_records
+        """).format(key=query_key,
+                    relation=relation)
+
+        record = self._execute_cypher_query(statement=query,
+                                            param_dict={})
+
+        if not record:
+            raise NotFoundException('User {user_id} does not {relation} '
+                                    'any resources'.format(user_id=user_id,
+                                                           relation=relation))
+        results = []
+        table_records = record.single().get('table_records', [])
+
+        for record in table_records:
+            # todo: decide whether we want to return a list of table entities or just table_uri
+            results.append(self.get_table(table_uri=record['key']))
+        return {'table': results}
+
+    @timer_with_counter
+    def add_resource_relation_by_user(self, *,
+                                      table_uri: str,
+                                      user: str,
+                                      relation: str,
+                                      reverse_relation: str) -> None:
+        """
+        Update table user informations.
+        1. Do a upsert of the user node.
+        2. Do a upsert of the relation/reverse-relation edge.
+
+        :param table_uri:
+        :param user:
+        :param relation:
+        :param reverse_relation:
+        :return:
+        """
+        upsert_user_query = textwrap.dedent("""
+        MERGE (u:User {key: $user_email})
+        on CREATE SET u={email: $user_email, key: $user_email}
+        on MATCH SET u={email: $user_email, key: $user_email}
+        """)
+
+        user_email = 'key: "{user_email}"'.format(user_email=user)
+        tbl_key = 'key: "{tbl_key}"'.format(tbl_key=table_uri)
+
+        upsert_user_relation_query = textwrap.dedent("""
+        MATCH (n1:User {{{user_email}}}), (n2:Table {{{tbl_key}}})
+        MERGE (n1)-[r1:{relation}]->(n2)-[r2:{reverse_relation}]->(n1)
+        RETURN n1.key, n2.key
+        """).format(user_email=user_email,
+                    tbl_key=tbl_key,
+                    relation=relation,
+                    reverse_relation=reverse_relation)
+
+        try:
+            tx = self._driver.session().begin_transaction()
+            # upsert the node
+            tx.run(upsert_user_query, {'user_email': user})
+            result = tx.run(upsert_user_relation_query, {})
+
+            if not result.single():
+                raise RuntimeError('Failed to create relation between '
+                                   'user {user} and table {tbl}'.format(user=user,
+                                                                        tbl=table_uri))
+        except Exception as e:
+            if not tx.closed():
+                tx.rollback()
+            # propagate the exception back to api
+            raise e
+        finally:
+            tx.commit()
```

Comment:
This will throw exception if it's rolled back shadowing original exception.
Could you refer to this one? https://github.com/lyft/amundsenmetadatalibrary/blob/master/metadata_service/proxy/neo4j_proxy.py#L319